consensus/blake3pow/consensus.go

```diff
@@ -0,0 +1,509 @@
+		alloc := core.ReadGenesisAlloc("genallocs/gen_alloc_" + common.NodeLocation.Name() + ".json")
+		log.Info("Allocating genesis accounts", "num", len(alloc))
+
+		for addressString, account := range alloc {
+			addr := common.HexToAddress(addressString)
+			internal, err := addr.InternalAddress()
+			if err != nil {
+				log.Error("Provided address in genesis block is out of scope")
+			}
+			state.AddBalance(internal, account.Balance)
+			state.SetCode(internal, account.Code)
+			state.SetNonce(internal, account.Nonce)
+			for key, value := range account.Storage {
+				state.SetState(internal, key, value)
+			}
+		}
+	}
+
+	header.SetRoot(state.IntermediateRoot(true))
+}
+
+// FinalizeAndAssemble implements consensus.Engine, accumulating the block and
+// uncle rewards, setting the final state and assembling the block.
+func (blake3pow *Blake3pow) FinalizeAndAssemble(chain consensus.ChainHeaderReader, header *types.Header, state *state.StateDB, txs []*types.Transaction, uncles []*types.Header, etxs []*types.Transaction, subManifest types.BlockManifest, receipts []*types.Receipt) (*types.Block, error) {
+	nodeCtx := common.NodeLocation.Context()
+	if nodeCtx == common.ZONE_CTX {
+		// Finalize block
+		blake3pow.Finalize(chain, header, state, txs, uncles)
+	}
+
+	// Header seems complete, assemble into a block and return
+	return types.NewBlock(header, txs, uncles, etxs, subManifest, receipts, trie.NewStackTrie(nil)), nil
+}
+
+func (blake3pow *Blake3pow) ComputePowLight(header *types.Header) (common.Hash, common.Hash) {
+	return common.Hash{}, common.Hash{}
+}
+
+// AccumulateRewards credits the coinbase of the given block with the mining
+// reward. The total reward consists of the static block reward and rewards for
+// included uncles. The coinbase of each uncle block is also rewarded.
+func accumulateRewards(config *params.ChainConfig, state *state.StateDB, header *types.Header, uncles []*types.Header) {
+	// Select the correct block reward based on chain progression
+	blockReward := misc.CalculateReward()
+
+	coinbase, err := header.Coinbase().InternalAddress()
+	if err != nil {
+		log.Error("Block has out of scope coinbase, skipping block reward", "Address", header.Coinbase().String(), "Hash", header.Hash().String())
+		return
+	}
+
+	// Accumulate the rewards for the miner and any included uncles
+	reward := new(big.Int).Set(blockReward)
+	r := new(big.Int)
+	for _, uncle := range uncles {
+		coinbase, err := uncle.Coinbase().InternalAddress()
+		if err != nil {
+			log.Error("Found uncle with out of scope coinbase, skipping reward", "Address", uncle.Coinbase().String(), "Hash", uncle.Hash().String())
+			continue
+		}
+		r.Add(uncle.Number(), big8)
+		r.Sub(r, header.Number())
+		r.Mul(r, blockReward)
+		r.Div(r, big8)
+		state.AddBalance(coinbase, r)
+
+		r.Div(blockReward, big32)
+		reward.Add(reward, r)
+	}
+	state.AddBalance(coinbase, reward)
+}
+
+func TargetToDifficulty(target *big.Int) *big.Int {
```

Comment:
`TargetToDifficulty()` and `DifficultyToTarget()` are already defined in `consensus/consensus.go`. Please don't redefine them here

Reply:
removed it completely